registry/cache/repocachemanager.go

```diff
@@ -0,0 +1,268 @@
+// It returns the images found, those which require updating and details about
+// why they need to be updated.
+func (c *repoCacheManager) fetchImages(tags []string) (fetchImagesResult, error) {
+	images := map[string]image.Info{}
+
+	// Create a list of images that need updating
+	var toUpdate []imageToUpdate
+
+	// Counters for reporting what happened
+	var missing, refresh int
+	for _, tag := range tags {
+		if tag == "" {
+			return fetchImagesResult{}, fmt.Errorf("empty tag in fetched tags")
+		}
+
+		// See if we have the manifest already cached
+		newID := c.repoID.ToRef(tag)
+		key := NewManifestKey(newID.CanonicalRef())
+		bytes, deadline, err := c.cacheClient.GetKey(key)
+		// If err, then we don't have it yet. Update.
+		switch {
+		case err != nil: // by and large these are cache misses, but any error shall count as "not found"
+			if err != ErrNotCached {
+				c.logger.Log("warning", "error from cache", "err", err, "ref", newID)
+			}
+			missing++
+			toUpdate = append(toUpdate, imageToUpdate{ref: newID, previousRefresh: initialRefresh})
+		case len(bytes) == 0:
+			c.logger.Log("warning", "empty result from cache", "ref", newID)
+			missing++
+			toUpdate = append(toUpdate, imageToUpdate{ref: newID, previousRefresh: initialRefresh})
+		default:
+			var entry registry.ImageEntry
+			if err := json.Unmarshal(bytes, &entry); err == nil {
+				if c.trace {
+					c.logger.Log("trace", "found cached manifest", "ref", newID, "last_fetched", entry.LastFetched.Format(time.RFC3339), "deadline", deadline.Format(time.RFC3339))
+				}
+
+				if entry.ExcludedReason == "" {
+					images[tag] = entry.Info
+					if c.now.After(deadline) {
+						previousRefresh := minRefresh
+						lastFetched := entry.Info.LastFetched
+						if !lastFetched.IsZero() {
+							previousRefresh = deadline.Sub(lastFetched)
+						}
+						toUpdate = append(toUpdate, imageToUpdate{ref: newID, previousRefresh: previousRefresh, previousDigest: entry.Info.Digest})
+						refresh++
+					}
+				} else {
+					if c.trace {
+						c.logger.Log("trace", "excluded in cache", "ref", newID, "reason", entry.ExcludedReason)
+					}
+					if c.now.After(deadline) {
+						toUpdate = append(toUpdate, imageToUpdate{ref: newID, previousRefresh: excludedRefresh})
+						refresh++
+					}
+				}
+			}
+		}
+	}
+
+	result := fetchImagesResult{
+		imagesFound:                images,
+		imagesToUpdate:             toUpdate,
+		imagesToUpdateRefreshCount: refresh,
+		imagesToUpdateMissingCount: missing,
+	}
+
+	return result, nil
+}
+
+// updateImages, refreshes the cache entries for the images passed. It may not succeed for all images.
+// It returns the values stored in cache, the number of images it succeeded for and the number
+// of images whose manifest wasn't found in the registry.
+func (c *repoCacheManager) updateImages(ctx context.Context, registryClient registry.Client, images []imageToUpdate) (map[string]image.Info, int, int) {
```

Comment:
Just a note; I am not a fan of the `.., int, int` return, but wrapping it in a struct to give it more context does not seem very meaningful either, the comment makes up for it a bit though.

Reply:
Yeah, I am not too pleased with `fetchImagesResult` either

Happy to look into any suggestions you may have.

Maybe computing the containers outside of the functions? I gave that a short chance but it was hairy. I will give it another try.

Reply:
I didn't find a cleaner way. At least not with a much bigger refactoring.